### PR TITLE
[Clang] Fix crash with `source_location` in lambda declarators.

### LINKED ISSUE #67134: Internal Compiler Error 17.0.1 and trunk with NTTP (16 sloc)
Minimal example on [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCAAzKQADqgKhE4MHt6%2BekkpjgJBIeEsUTHxdpgOaUIETMQEGT5%2BXLaY9nkM1bUEBWGR0XG2NXUNWc0KQ93BvcX9sQCUtqhexMjsHOaxwcjeWADUJrFuCksrmAD6tKii7QfYJhoAghtbO5j7h8jj6FhUt/dPjwImBYCQMQPebgIAE8EoxWG83AhaqRdl8QCAUgAvc4EXahP6PcbELwOXZsVAAd32AHYrI9drs0AxxpgAG5iMmYSkQRG1RkCca7CDmABsRLmJgArBZ8ZKACIotEYk6rC5XJjtXaXZDvOWogjodHHZaq7UatLo5DLYiMAgQOZzXYMVCYVSrBK4ky0r1y/4%2Bg50gEPIEgsFvA5uckUgkPIkk3F4VY0qzU32xQP/YK0Kb85lA1QJYi7JheIi7VS6/bSqVy%2B27AC0t12icwEfMZmpIvbTa9KbTGcewVxLCYwXtvf%2BDIZNoIywYuw0Ab9qY4C1onElvD8HC0pFQnDc1msqJV4bMsR4pAImlXCwA1iAzJIAHQXgCcIrMIrfAA432YNB/SUf3idcOEkXgWAkDQNFIbdd33DheAUEBYOvHdV1IOBYBgRAUFQEE6GichKDQQj6BibZDGAM4CGJBg71ILBWRbAA1PBMApAB5WFt0vGhaCBYgUIgCIb1ICJglqKFOEvSTmGIKEuIibQKnQy8yLYQQuIYWgZIwpjMAiLxgDcMRaBQ7heCwEcjHEAz8BtSpWUwSzd1dCpSzWXch1acTswiYhpI8LBZN4Oi8CgqyFioAxgAUdjOJ4xgwpkQQRDEdgpDS%2BQlDUcTdDMfRqJQI9LH0PAIhQyAFlQD00ks%2BsvgOOVTEsawAIbL4ApagK2qsSwGy4swGw9BAbSYdBeFQFziGIPAsGq%2B0WjaC0IFcEYmlIQIpiKEpsmSVIBE2g7cjSHo9v6MZWjUqoJhO67VoETo6guvoYjGe7PEaPRxi6N6Zg%2BhZjVOCQ1w3LdxMQ8sfxFesRUkRkDCMIU6K8BjHQgXBCBIfZzy4OZwpvB1SAfWIX3PMxYa4WJv2pP9yf0TgINIKDJVg%2BDps4ZDUKvYmsNwiAkCWAgElLEj1vwIh5sNZp%2BHS0RxGy%2BXcpUdQDMK0gKSChIwvBjhNzgqHOC40sxdxVAqBhuGEaR6jUfou9MY8cjojx2IzEJvmMJJhBMEm/plrAlm2Y542kNsXn0K0EmHy4H9n0/H9qQ/ZOEepLhM6ZjhYkhgzEKJn2BfgbC8LIhIiLICgIHLyuUGR4AuBFGDDJY1ZEu43jUoEoSRLEgz5Ok1LB8U5TVIcVLNNtHS9PErBjNM8zLMvGzqPs3dHNuly3N4DyrSBcTfLA3cAqCxSQu88L5qingYrihKOM7lKrJyjKlekFXFDVgq9Ab0r2vKgFJatV6oCk4E1A0LV%2BodQ0ENEa9YxoTSmnuWaMtXLwGBjdSozh1oMHcN9UY208EA32s0HIR10gEK2uQ9oJCrorVus9L6mQtrlGwR0CYdCPqDC6A9Hhr1drvTBosE0WV9aG05nuTg1t4aIyoijJuz4NBKKFNjaW7sCaFxjgsP2AcYjLTjiKV8b5qSxB/H%2BSUUgOwihAtnEOIB2ZG3ztzSOaF%2BalyFiAEW5sJaYCliQBaegVZvyyh/WQX98oaz0NrJgutorZwkeHU2otSy7EtjI228jgC7EUco2BPICIVwohor20dbykF0VgfR%2Bt7GOMkQXVx3ttGkwkNSZ81iuD/jMJTSU55QKcFzk4hCLiykLA8YLJAtcKISymf0eKzAEgKAQKgAgrc2KP2SnxXgPdoh93EiPfSckpKjxUmpSeBEtIEBnvpDeRkTJmVoBZVKq87KXyYngJyjht7iT3l5Q%2Bgg/IGVPsFDAbyIo30wrFJg8UO6bO7rIEJEgwmCAierXcuhQIlWgYAyqwC9ygOZOA5qsRWplQsABaaqCFroJqgw9hLg8F8J2oUIRZDDrtD4TQ86gjAa/Swe0F69QqF8qehw/6PLSH8KFSw36nCJVXWBqeYRYFEnOI4BkxGwBkA6ggGjDGqj/FFg2Jopp5TKmB3vI%2BWIz5qTUkkFwQCAETH/jfBofp4FIIOLDmqnmbii4eOFmbcW1dZnEFCPCTgqhYayLtgooxeTd6GsCXLBFitQk5VRT/eIMS4m3wSXnYZHBknmzSVbKNNtNXaodujJ2QoXZFLdhseYWizX%2ByqZQGpnq6nh19aa2OEg3w2upJKMxsQ3yZw0DY7Ogz6kjPcYLCZ%2BFXZV1IoUuuWSuDUg0M0Zi6ykpdxfjs4SlB%2B67gOcPY5SlTkTxflPbSukbnWTuYvR5y8n22WAOvayHyt7oIMr8g%2BL8j7%2BUqmfKEF9xLgr1nwe%2BsKD38VTZlJFGa8pop0CAaQf9sU2CARg/F7RGrEtJQA8lCEqWLTw2w9oDL8EytdUQ9AXCMOJHZWkE69GuX5HlTEaQVG7q8OFfRvjTDxUst5bx5hP0hNyrE/tSQirRHKohkMrm6ry0xqyTktpyiuAGpxka/GpTiaWoAkomC5mLMWcw8zLt3rC29tGcXWAgaUmrJDWuii4a2CRujZkhuWmlHPkvH4/TybX5puQ5/VDP9pA5ugyqgtqni2pPSepvz9tN2Bd0wU5d7t5Mtt9m2i1djbMqakRHFCfrmkPlM5Zur5nrM50S%2BVgr%2BszDNYaaMhYs0UjOEkEAA%3D%3D%3D).

### PR DESCRIPTION
**Pull request by cor3ntin**
Parsing lambdas require pushing a declaration context for the lambda, so that parameters can be attached to it, before its trailing type is parsed. DAt that point, partially-parsed lambda don't have a name that can be computed for then.
This would cause source_location::current() to crash when use in the decltype of a lambda().
We work around this by producing a source_location for an enclosing scope in that scenario.

Fixes #67134

### CODE REVIEW

File: clang/lib/AST/Expr.cpp [hunk 1]
```diff
@@ -2287,6 +2288,15 @@ APValue SourceLocExpr::EvaluateInContext(const ASTContext &Ctx,
...
   }
 
+  // If we are currently parsing a lambda declarator, we might not have a fully
+  // formed call operator declaration yet, and we could not form a function name
+  // for it. Because we do not have access to Sema/function scopes here, we
+  // detect this case by relying on the fact such method doesn't yet have a
+  // type.
+  if (const auto *D = dyn_cast<CXXMethodDecl>(Context);
+      D && D->getFunctionTypeLoc().isNull() && isLambdaCallOperator(D))
+    Context = D->getParent()->getParent();
```
Comment (erichkeane) [inline clang/lib/AST/Expr.cpp:2298]: So the first `getParent` should get the lambda object, correct?  

Could this cause issues where the parent of the lambda is a TU?  Can we have a test for that?  Also, there are some goofinesses around variable templates and requires in a lambda, can we have a test for that?

**Review by erichkeane — APPROVED** [review]

**Comment by erichkeane** [comment]
Woops!  Had the wrong tab open and typed in the wrong 'fixes', did not mean to close this.

The patch was supposed to fix https://github.com/llvm/llvm-project/issues/107052

**Comment by zyn0217** [comment]
While looking at it, I found another bug where we seem to have problems mangling `std::source_location`
https://clang.godbolt.org/z/467E19h45

File: clang/lib/AST/Expr.cpp (hunk 1 above)
Comment (zyn0217) [reply clang/lib/AST/Expr.cpp:2298]: Do we need to make sure we're in an unevaluated context here?

**Comment by cor3ntin** [comment]
> While looking at it, I found another bug where we seem to have problems mangling `std::source_location` https://clang.godbolt.org/z/467E19h45

See https://github.com/llvm/llvm-project/issues/59948#issuecomment-2331676234 :)

File: clang/lib/AST/Expr.cpp (hunk 1 above)
Comment (cor3ntin) [reply clang/lib/AST/Expr.cpp:2298]: ideally yes, I'd assert on it... but we do not have access to that information in constant evaluation

**Pull request merged**